### PR TITLE
DRILL-6255: Drillbit while sending control message to itself creates …

### PR DESCRIPTION
**Pull request by sohami**
…a connection instead of submitting locally

### CODE REVIEW

**Comment by sohami** [comment]
@arina-ielchiieva - Please help to review this PR.

File: exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/FragmentsRunner.java [hunk 1]
```diff
@@ -180,44 +170,38 @@ private void setupNonRootFragments(final Collection<PlanFragment> fragments) thr
...
-    for (final PlanFragment fragment : localLeafFragmentList) {
-      startLocalFragment(fragment);
+    for (final DrillbitEndpoint ep : leafFragmentMap.keySet()) {
+      sendRemoteFragments(ep, leafFragmentMap.get(ep), null, null);
     }
   }
 
   /**
-   * Send all the remote fragments belonging to a single target drillbit in one request.
+   * Send all the remote fragments belonging to a single target drillbit in one request. If the assignment
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/FragmentsRunner.java]: I think we should consider renaming: `sendRemoteFragments` -> `sendLeafFragments`, `scheduleRemoteIntermediateFragments`  -> `scheduleIntermediateFragments`.

File: exec/java-exec/src/main/java/org/apache/drill/exec/work/batch/ControlMessageHandler.java [hunk 2]
```diff
@@ -254,6 +249,32 @@ private Ack receivingFragmentFinished(final FinishedReceiver finishedReceiver) {
...
+    return Acks.OK;
+  }
+
+  public Ack requestQueryCancel(QueryId queryId) {
+    return bee.cancelForeman(queryId, null) ? Acks.OK : Acks.FAIL;
+  }
+
+  public Ack initializeFragment(InitializeFragments fragments) throws RpcException {
+    final DrillbitContext drillbitContext = bee.getContext();
+    for(int i = 0; i < fragments.getFragmentCount(); i++) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/work/batch/ControlMessageHandler.java]: Space after for.

File: exec/java-exec/src/main/java/org/apache/drill/exec/rpc/control/LocalControlConnectionManager.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+   * @throws OutOfMemoryException
+   */
+  private DrillBuf convertToByteBuf(ByteBuf[] byteBuffArray) throws OutOfMemoryException {
+
+    if (byteBuffArray == null) {
+      return null;
+    }
+
+    int bytesToCopy = 0;
+    for(ByteBuf b : byteBuffArray) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/rpc/control/LocalControlConnectionManager.java]: Space after for.

File: exec/java-exec/src/main/java/org/apache/drill/exec/rpc/control/LocalControlConnectionManager.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+      final int validBytes = b.readableBytes();
+      if (0 == validBytes) {
+        b.release();
+      } else {
+        bytesToCopy += validBytes;
+      }
+    }
+    final DrillBuf drillBuff = config.getAllocator().buffer(bytesToCopy);
+
+    for(ByteBuf b : byteBuffArray) {
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/rpc/control/LocalControlConnectionManager.java]: Space after for.

File: exec/java-exec/src/main/java/org/apache/drill/exec/rpc/control/LocalControlConnectionManager.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+      case BitControl.RpcType.REQ_UNPAUSE_FRAGMENT_VALUE: {
+        final ControlTunnel.SignalFragment signalFragment = ((ControlTunnel.SignalFragment) cmd);
+        final RpcOutcomeListener<Ack> outcomeListener = signalFragment.getOutcomeListener();
+        final Ack ackResponse = messageHandler.resumeFragment(signalFragment.getMessage());
+        outcomeListener.success(ackResponse, null);
+        break;
+      }
+
+      default:
+        final RpcException rpcException = new RpcException(String.format("Unsupported control request type {} " +
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/main/java/org/apache/drill/exec/rpc/control/LocalControlConnectionManager.java]: Replace `{}` with `%s`.

**Comment by sohami** [comment]
Rebased on top of latest master

**Comment by ilooner** [comment]
@sohami Please fix Travis failures

```
Tests in error: 
  TestLocalControlConnectionManager.setupForTest:101 WrongTypeOfReturnValue 
Con...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
  TestLocalControlConnectionManager.setupForTest:101 MissingMethodInvocation 
wh...
```

File: exec/java-exec/src/test/java/org/apache/drill/exec/rpc/control/TestLocalControlConnectionManager.java [hunk 6]
```diff
@@ -55,6 +58,8 @@
 
   private CountDownLatch latch;
 
+  private final String negativeAckMessage = "Negative Ack received";
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/rpc/control/TestLocalControlConnectionManager.java]: static? and `NEGATIVE_ACK_MESSAGE`.

File: exec/java-exec/src/test/java/org/apache/drill/exec/rpc/control/ConnectionManagerRegistryTest.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+
+public class ConnectionManagerRegistryTest {
+
+  private final DrillbitEndpoint localEndpoint = DrillbitEndpoint.newBuilder()
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/rpc/control/ConnectionManagerRegistryTest.java]: static here and for `foremanEndpoint`.

File: exec/java-exec/src/test/java/org/apache/drill/exec/rpc/control/TestLocalControlConnectionManager.java [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+  private static ControlConnectionConfig mockConfig;
+
+  private static ControlMessageHandler mockHandler;
+
+  private static ControlTunnel controlTunnel;
+
+  private CountDownLatch latch;
+
+  private final String negativeAckMessage = "Negative Ack received";
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/rpc/control/TestLocalControlConnectionManager.java]: static

File: exec/java-exec/src/test/java/org/apache/drill/exec/rpc/control/TestLocalControlConnectionManager.java [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+  private final RpcOutcomeListener<GeneralRPCProtos.Ack> outcomeListener =
```
Comment (arina-ielchiieva) [inline exec/java-exec/src/test/java/org/apache/drill/exec/rpc/control/TestLocalControlConnectionManager.java]: static

**Comment by sohami** [comment]
@arina-ielchiieva - Made all the changes, rebased and squashed the commits.

**Comment by arina-ielchiieva** [comment]
+1, LGTM.

**Pull request closed without merging**